packages/pool/src/weighted.rs

```diff
@@ -0,0 +1,478 @@
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{Coin, Decimal, StdResult, Uint256};
+use cosmwasm_std::{Decimal256, Deps, StdError};
+use ibcx_math::MathError;
+
+use crate::PoolError;
+
+use super::OsmosisPool;
+
+fn to_dec(v: Uint256) -> Result<Decimal256, PoolError> {
+    Ok(Decimal256::from_atomics(v, 0)?)
+}
+
+fn solve_constant_function_invariants(
+    token_balance_fixed_before: Decimal256,
+    token_balance_fixed_after: Decimal256,
+    token_weight_fixed: Decimal256,
+    token_balance_unknown_before: Decimal256,
+    token_weight_unknown: Decimal256,
+) -> Result<(Decimal256, bool), MathError> {
+    let weight_ratio = token_weight_fixed / token_weight_unknown;
+
+    let y = token_balance_fixed_before / token_balance_fixed_after;
+
+    let y_to_weight_ratio = ibcx_math::pow(y, weight_ratio)?;
+
+    let (paranthetical, neg) = ibcx_math::abs_diff_with_sign(Decimal256::one(), y_to_weight_ratio);
+
+    let amount_y = token_balance_unknown_before * paranthetical;
+
+    Ok((amount_y, neg))
+}
+
+#[cw_serde]
+pub struct Pool {
+    #[serde(rename = "@type")]
+    pub type_url: String,
+    pub address: String,
+    pub id: String,
+    pub future_pool_governor: String,
+    pub pool_params: PoolParams,
+    pub pool_assets: Vec<PoolAsset>,
+    pub total_shares: Coin,
+    pub total_weight: Uint256,
+}
+
+#[cw_serde]
+pub struct PoolParams {
+    pub swap_fee: Decimal,
+    pub exit_fee: Decimal,
+    pub smooth_weight_change_params: Option<PoolSmoothWeightChangeParams>,
+}
+
+#[cw_serde]
+pub struct BigCoin {
+    pub denom: String,
+    pub amount: Uint256,
+}
+
+#[cw_serde]
+pub struct PoolAsset {
+    pub token: BigCoin,
+    pub weight: Uint256,
+}
+
+#[cw_serde]
+pub struct PoolSmoothWeightChangeParams {
+    pub initial_pool_weights: Vec<PoolAsset>,
+    pub target_pool_weights: Vec<PoolAsset>,
+    pub start_time: String,
+    pub duration: String,
+}
+
+struct PoolAssetTuple(pub (String, Uint256, Uint256));
+
+impl From<PoolAssetTuple> for PoolAsset {
+    fn from(v: PoolAssetTuple) -> Self {
+        Self {
+            token: BigCoin {
+                denom: v.0 .0,
+                amount: v.0 .1,
+            },
+            weight: v.0 .2,
+        }
+    }
+}
+
+impl Pool {
+    fn get_asset(&self, denom: &str) -> Result<PoolAsset, PoolError> {
+        Ok(self
+            .pool_assets
+            .iter()
+            .find(|v| v.token.denom == denom)
+            .ok_or_else(|| StdError::generic_err(format!("asset {denom} not found")))?
+            .clone())
+    }
+
+    fn apply_new_pool_assets(
+        &mut self,
+        input_denom: &str,
+        output_denom: &str,
+        input_value: Uint256,
+        output_value: Uint256,
+    ) -> Result<(), PoolError> {
+        let pool_assets = self.pool_assets.clone();
```

Comment:
Can we modify original vector in place to save memory?

Reply:
We can, will apply it